keel-tagging-plugin/src/main/kotlin/com/netflix/spinnaker/keel/tagging/KeelTagHandler.kt

```diff
@@ -0,0 +1,189 @@
+/*
+ *
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.netflix.spinnaker.keel.tagging
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ResourceKind
+import com.netflix.spinnaker.keel.api.ResourceName
+import com.netflix.spinnaker.keel.api.SPINNAKER_API_V1
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.events.TaskRef
+import com.netflix.spinnaker.keel.model.Job
+import com.netflix.spinnaker.keel.model.OrchestrationRequest
+import com.netflix.spinnaker.keel.model.OrchestrationTrigger
+import com.netflix.spinnaker.keel.orca.OrcaService
+import com.netflix.spinnaker.keel.plugin.ResolvableResourceHandler
+import com.netflix.spinnaker.keel.plugin.ResourceDiff
+import com.netflix.spinnaker.keel.plugin.ResourceNormalizer
+import com.netflix.spinnaker.keel.retrofit.isNotFound
+import com.netflix.spinnaker.keel.tags.EntityTag
+import com.netflix.spinnaker.keel.tags.EntityTags
+import com.netflix.spinnaker.keel.tags.KEEL_TAG_NAME
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import retrofit2.HttpException
+
+/**
+ * Ensures that keel resources are tagged with entity tags indicating
+ * management by keel
+ */
+class KeelTagHandler(
+  private val cloudDriverService: CloudDriverService,
+  private val orcaService: OrcaService,
+  override val objectMapper: ObjectMapper,
+  override val normalizers: List<ResourceNormalizer<*>>
+) : ResolvableResourceHandler<KeelTagSpec, TaggedResource> {
+
+  override val log: Logger by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override val apiVersion = SPINNAKER_API_V1.subApi("tag")
+  override val supportedKind = ResourceKind(
+    apiVersion.group,
+    "keel-tag",
+    "keel-tags"
+  ) to KeelTagSpec::class.java
+
+  override fun generateName(spec: KeelTagSpec) = ResourceName("tag:keel-tag:${spec.keelId}")
+
+  override suspend fun desired(resource: Resource<KeelTagSpec>): TaggedResource {
+    when (resource.spec.tagState) {
+      is TagDesired -> {
+        val desiredTag = (resource.spec.tagState as TagDesired).tag
+        return TaggedResource(resource.spec.keelId, resource.spec.entityRef, desiredTag)
+      }
+      else -> {
```

Comment:
If `TagState` is a sealed class you should be able to specify both options and not need an else.